src/geometry.cpp

```diff
@@ -107,5 +111,87 @@ namespace geometry {
 
 	}
 
+	std::vector<Cell> polygonOutlineCells(const std::vector<Cell>& polygon, int size_x) {	  
+	  // The output vector of cells
+	  std::vector<Cell> out;	  
+
+	  // If the input is less than or equal to one element, return the input
+	  if (polygon.size() <= 1) {
+	  	return polygon;
+	  }
+
+	  // Iterator to the next vertex
+	  auto next_vrtx = polygon.begin();
+
+	  for (const auto &vertex:polygon) {
+	  	// If the next vertex is not the end iterator, advance the next_vertex to point to the next iterator
+	  	if (next_vrtx != polygon.end()) {
+	  		std::advance(next_vrtx,1);
+	  	} 
+	  	// After advancing, if the iterator points to the end, go back to the beginning, essentially closing the polygon
+	  	if (next_vrtx == polygon.end()) {
+	  		next_vrtx = polygon.begin();
+	  	}
+
+	  	// Generate the pixels from one vertex to the next
+	  	auto temp = raytrace(vertex, (*next_vrtx), size_x);
+
+	  	// Remove the last element of the vector because the next vertex will use it as a starting point i.e. don't duplicate pixels
+	  	temp.pop_back();
+
+	  	// Push the pixels into the out vector
+	  	for (const auto &pixel:temp) {
+	  		out.push_back(pixel);
+	  	}
+
+	  }
+	  return out;
+	}	
+
+	std::vector<Cell> convexFillCells(const int size_x, const std::vector<Cell>& polygon) {
+	  // we need a minimum polygon of a triangle (Should this return the polygon?)
+	  if (polygon.size() < 3) return polygon;
+
+	  const auto polygon_outline = polygonOutlineCells(polygon, size_x);
+
+	  std::unordered_map<int,std::pair<int,int>> x_and_extrema;
+
+	  //TODO: turn the nested if else to else ifs
+
+	  // Find all the x values and extrema
+	  for (const auto pixel:polygon_outline) {
+	  	// Check if the x value for the pixel is in x_and_extrema
+	  	if (x_and_extrema.find(pixel.x) == x_and_extrema.end()) {
+	  		// if it isnt in x_and_extrema push the x in and with the y value as both the min and the max
+	  		x_and_extrema[pixel.x] = {pixel.y, pixel.y};
+	  	} else {
+	  		// if it is in x_and_extrema, check the y value. If the y value is lower than the minima,
+	  		// replace the first value, else if it is greater than the maxima, replace the second 
+	  		// value, else do nothing
+	  		if (pixel.y < ((x_and_extrema.at(pixel.x)).first)) {
+	  			(x_and_extrema.at(pixel.x)).first = pixel.y;
+	  		} else if (pixel.y > ((x_and_extrema.at(pixel.x)).second)) {
+	  			(x_and_extrema.at(pixel.x)).second = pixel.y;
+	  		}
+	  	}
+	  }
+
+	  std::vector<Cell> out;
+
+	  // lambda function
+	  auto fill_cells = [&out](const std::pair<int,std::pair<int,int>> &x_and_lims){
+	  	for (auto lower_bound = (x_and_lims.second).first; lower_bound <= (x_and_lims.second).second; lower_bound++) {
+	  		out.push_back({x_and_lims.first, lower_bound});
```

Comment:
Can we try something like
```
std::for_each(extrema.begin(), extrema.end(), [&out](auto& extremum){
  const int n = extremum.second.second - extremum.second.first + 1;
  std::generate_n(std::back_inserter(out), n, [extremum]() mutable {
    return Cell{extremum.first, extremum.second.first++};
  });
});
```
https://godbolt.org/z/Kbx43n

Reply:
Would be nice though if things were named instead of `second.first` :stuck_out_tongue:

Reply:
He? No? No like?